error.go

```diff
@@ -131,17 +131,24 @@ type Error struct {
 	Err error
 	// ResultCode is the LDAP error code
 	ResultCode uint8
+	// MatchedDN is the matchedDN returned in the LDAP error
+	MatchedDN string
 }
 
 func (e *Error) Error() string {
-	return fmt.Sprintf("LDAP Result Code %d %q: %s", e.ResultCode, LDAPResultCodeMap[e.ResultCode], e.Err.Error())
+	return fmt.Sprintf("LDAP Result Code %d %q: %s matchedDN: %s", e.ResultCode, LDAPResultCodeMap[e.ResultCode], e.Err.Error(), e.MatchedDN)
 }
 
 // NewError creates an LDAP error with the given code and underlying error
 func NewError(resultCode uint8, err error) error {
```

Comment:
I would just change this and drop `NewErrorWithDN()` to have this PR only go into the upcoming v3... @liggitt what do you think?

Reply:
The `Error` structure, gets a MatchedDN populated, only if the values returned by `getLDAPResultCode()` is put into it. If you drop `NewErrorWithDN()`, then `NewError()` would have to take 3 arguments.  This would make code blocks like:

    packetResponse, ok := <-msgCtx.responses
    if !ok {
        return NewError(ErrorNetwork, errors.New("ldap: response channel closed")
    }

Needing an empty string argument for MatchedDN. 

Can we make `getLDAPResultCode` into something like `GetLDAPError` which will return the full `Error` object and do a `err != nil` check. So places like this:

    if packet.Children[1].Tag == ApplicationAddResponse {
        resultCode, resultDescription, matchedDN := getLDAPResultCode(packet)
        if resultCode != 0 {
            return NewErrorWithDN(resultCode, errors.New(resultDescription), matchedDN)
        }
     }

Would become

    if packet.Children[1].Tag == ApplicationAddResponse {
        err := GetLDAPError(packet)
        if err != nil {
            return err
        }
     }

Then we can retain `NewError` as is, and do away with `NewErrorWithDN()`.

Reply:
I like this suggestion - do you want to update this PR accordingly?